src/ddmd/e2ir.d

```diff
@@ -5794,9 +5793,25 @@ private elem *appendDtors(IRState *irs, elem *er, size_t starti, size_t endi)
 elem *toElemDtor(Expression e, IRState *irs)
 {
     //printf("Expression.toElemDtor() %s\n", e.toChars());
-    const starti = irs.varsInScope ? irs.varsInScope.dim : 0;
+
+    /* "may" throw may actually be false if we look at a subset of
+     * the function. Here, the subset is `e`. If that subset is nothrow,
+     * we can generate much better code for the destructors for that subset,
+     * even if the rest of the function throws.
+     * If mayThrow is false, it cannot be true for some subset of the function,
+     * so no need to check.
+     * If calling canThrow() here turns out to be too expensive,
+     * it can be enabled only for optimized builds.
+     */
+    const mayThrowSave = irs.mayThrow;
+    if (irs.mayThrow && !canThrow(e, irs.getFunc(), false))
+        irs.mayThrow = false;
+
+    const starti = irs.varsInScope.dim;
     elem* er = toElem(e, irs);
-    const endi = irs.varsInScope ? irs.varsInScope.dim : 0;
+    const endi = irs.varsInScope.dim;
+
+    irs.mayThrow = mayThrowSave;
```

Comment:
Can you move the declarations for `starti` and `endi` below this line, as they're not related to determining if the expressing may throw, as far as I can see.

Reply:
They have to bracket the call to `toElem()`.

Reply:
Yes, you're right - I see that I overlooked the code. Sorry for the noise.